Mathlib/Algebra/Category/ModuleCat/OverFilteredColimits.lean

```diff
@@ -0,0 +1,351 @@
+import Mathlib.CategoryTheory.Limits.ConcreteCategory
+import Mathlib.CategoryTheory.ConcreteCategory.Elementwise
+
+/-!
+# Module structures of filtered colimits of abelian groups over filtered colimts of rings
+
+Let `R` be the filtered colimit of rings `{Rⱼ}` and `M` be the filtered colimit of
+abelian groups `{Mⱼ}`  with the same indexing set `j ∈ J`, if for each `j ∈ J`, `Mⱼ` is an `Rⱼ` such
+that the `Rⱼ`-action is compatible, then `M` is an `Rⱼ`-module for all `j` and `M` is an `R`-module.
+
+## Implementation notes
+
+We choose not to use `PresheafOfModules` to avoid code duplication:
+consider `R : J ⥤ CommRingCat` and `M : J ⥤ AddCommGrp`, then `colimit M` is both a
+`colimit R`-module and a `colimt (R ⋙ forget₂ CommRingCat RingCat)`-module; the two module
+structures are virtually the same. This situation manifests in stalks of sheaves of modules:
+for any ringed space `X` and a sheaf of `𝒪_X`-module `ℳ`, we want to think the stalk `ℳₓ` as an
+`𝒪_{X,x}`-module. But since `PresheafOfModules` requires a presheaf of `RingCat` not `CommRingCat`,
+we need to compose the sheaf with forgetful functors, but we don't want to think about the
+difference between `𝒪_{X, x}` as a colimit in `CommRing` and `𝒪_{X, x}` as a colimit in `RingCat`
+all the time. So we ask `R` and `M` to be functors into concrete categories which behaves like rings
+and abelian groups respectively.
+
+-/
+
+open CategoryTheory Category Limits Opposite
+
+universe u u' v v' w
+
+section
+
+variable {J : Type w} [Category J] [IsFiltered J]
+variable {ℜ𝔦𝔫𝔤 : Type u} [Category.{u'} ℜ𝔦𝔫𝔤] [ConcreteCategory.{w} ℜ𝔦𝔫𝔤]
+variable {𝔄𝔟 : Type v} [Category.{v'} 𝔄𝔟] [ConcreteCategory.{w} 𝔄𝔟]
+
+attribute [local instance] ConcreteCategory.hasCoeToSort
+attribute [local instance] ConcreteCategory.instFunLike
+
+variable [∀ x : ℜ𝔦𝔫𝔤, Semiring x] [∀ x : 𝔄𝔟, AddCommMonoid x]
+variable [∀ x y : ℜ𝔦𝔫𝔤, RingHomClass (x ⟶ y) x y]
+variable [∀ x y : 𝔄𝔟, AddMonoidHomClass (x ⟶ y) x y]
+
+variable (ℛ : J ⥤ ℜ𝔦𝔫𝔤) (ℳ : J ⥤ 𝔄𝔟)
+variable [HasColimit ℛ] [HasColimit ℳ]
+variable [PreservesColimit ℛ (forget ℜ𝔦𝔫𝔤)] [PreservesColimit ℳ (forget 𝔄𝔟)]
+
+variable [∀ c, Module (ℛ.obj c) (ℳ.obj c)]
+variable [compatible_smul : Fact $ ∀ {c₁ c₂ : J} (i₁ : c₁ ⟶ c₂) (r : ℛ.obj c₁) (m : ℳ.obj c₁),
+    ℳ.map i₁ (r • m) = ℛ.map i₁ r • ℳ.map i₁ m]
+
+namespace Module.overFilteredColimits
+
+variable {ℛ ℳ} in
+/--
+Let `R` be the filtered colimit of rings `{Rⱼ}` and `M` be the filtered colimit of
+abelian groups `{Mⱼ}`  with the same indexing set `j ∈ J`, if for each `j ∈ J`, `Mⱼ` is an `Rⱼ` such
+that the `Rⱼ`-action is compatible, then there is a heterogeneous scalar multiplication
+`Rᵢ → Mⱼ → Mₖ` for every `i → j` and `i → k`.
+-/
+def hsmul {c₁ c₂ c₃ : J} (i₁ : c₁ ⟶ c₃) (i₂ : c₂ ⟶ c₃)
+    (r : ℛ.obj c₁) (m : ℳ.obj c₂) : ℳ.obj c₃ :=
+  (ℛ.map i₁ r) • (ℳ.map i₂ m)
+
+namespace hsmul
+
+variable {c₁ c₂ c₃ : J} (i₁ : c₁ ⟶ c₃) (i₂ : c₂ ⟶ c₃)
+variable (r : ℛ.obj c₁) (m : ℳ.obj c₂)
+
+protected lemma one_smul :
```

Comment:
Should this be `one_hsmul` and not in an hsmul namespace at all? You're already in this overFilteredColimits namespace.